onnx/defs/math/defs.cc

```diff
@@ -1226,20 +1236,144 @@ Computes the error function of the given input tensor element-wise.
+        .Input(3, "b", "N-dimensional quantized matrix b", "T2")
+        .Input(4, "b_scale", "scale of quantized input b", "tensor(float)")
+        .Input(5, "b_zero_point", "zero point of quantized input b", "T2")
+        .Input(6, "y_scale", "scale of quantized output y", "tensor(float)")
+        .Input(7, "y_zero_point", "zero point of quantized output y", "T3")
+        .Output(0, "y", "Quantized matrix multiply results from a * b", "T3")
+        .TypeConstraint(
+            "T1",
+            {"tensor(int8)", "tensor(uint8)"},
+            "Constrain input a and its zero point data type to 8-bit integer tensor.")
+        .TypeConstraint(
+            "T2",
+            {"tensor(int8)", "tensor(uint8)"},
+            "Constrain input b and its zero point data type to 8-bit integer tensor.")
+        .TypeConstraint(
+            "T3",
+            {"tensor(int8)", "tensor(uint8)"},
+            "Constrain output y and its zero point data type to 8-bit integer tensor.")
+        .TypeAndShapeInferenceFunction([](ONNX_NAMESPACE::InferenceContext&
+                                              ctx) {
+          auto a_type = ctx.getInputType(0);
+          auto b_type = ctx.getInputType(3);
+          auto y_type = ctx.getOutputType(0);
+          if (nullptr == a_type || nullptr == b_type || nullptr == y_type ||
+              a_type->value_case() != ONNX_NAMESPACE::TypeProto::kTensorType ||
+              b_type->value_case() != ONNX_NAMESPACE::TypeProto::kTensorType) {
+            fail_type_inference(
+                "inputs are expected to have tensor type and output type should not be null.");
+          }
+
+          if (ONNX_NAMESPACE::TensorProto::UINT8 ==
+                  a_type->tensor_type().elem_type() &&
+              ONNX_NAMESPACE::TensorProto::UINT8 ==
+                  b_type->tensor_type().elem_type()) {
+            y_type->mutable_tensor_type()->set_elem_type(
+                ONNX_NAMESPACE::TensorProto::UINT8);
+          } else {
+            y_type->mutable_tensor_type()->set_elem_type(
+                ONNX_NAMESPACE::TensorProto::INT8);
+          }
+
+          matmulShapeInference(ctx, 0, 3);
+        }));
+
+static const char* MatMulInteger_ver10_doc = R"DOC(
+Matrix product that behaves like numpy.matmul: https://docs.scipy.org/doc/numpy-1.13.0/reference/generated/numpy.matmul.html.
+The production MUST never overflow. The accumulation may overflow if and only if in 32 bits.
+)DOC";
+
+ONNX_OPERATOR_SET_SCHEMA(
+    MatMulInteger,
+    10,
+    OpSchema()
+        .SetDoc(MatMulInteger_ver10_doc)
+        .Input(0, "A", "N-dimensional matrix A", "T1")
+        .Input(1, "B", "N-dimensional matrix B", "T2")
+        .Input(
+            2,
+            "a_zero_point",
+            "Zero point tensor for input 'A'. It's optional and default value is 0. It could be a scalar or a 1-D tensor, "
+            "which means a per-tensor or per-row quantization. If it's a 1-D tensor, its number of elements "
```

Comment:
We should explicitly pass the axis as an argument here to be consistent with the definition of per-row/per-column/per-channel quantization.

Reply:
I don't think that we need this flexibility, since for per-channel, it has to be per output channel (for conv) for weights. For matmul, if it's not per tensor, per row for the first and per-column for the 2nd is a good one, but per-column for the 1st input and per-row for 2nd input is not that useful (math is not that straight-forward). So I'd suggest to keep this specification.

Reply:
I think for matmul, per-row quant matmul per-column quant makes sense in terms of math. the other way (per-column quant matmul per-row quant) does not make sense, right?

For Conv, normally, folks just use per-tensor for activation and per output channel for weights some times. Don't really see other use cases. 

To not add complexity/burden to the spec and partner's implementation, I'd keep the spec as it is, and we may extend it when seeing more other cases.

Sounds good?

Reply:
I see what your saying but it might be good to be consistent across the spec. For Quantize_Linear, we specify axis as attribute:
   .Attr(
            "axis",
            "The axis along which same quantization parameters are applied. It's optional. If it's not specified, it means per-tensor quantization and input 'x_scale' and 'x_zero_point' must be scalars. If it's specified, it means per 'axis' quantization and input 'x_scale' and 'x_zero_point' must be 1-D tensors.",
            AttributeProto::INT,
            false)

Might be good to make this consistent across all ops. We could still enforce that the op itself only supports per-row for the first tensor and per column for the second one.

Reply:
My bad. "axis" for quant should be removed.